apollo-router-core/src/query_planner/mod.rs

```diff
@@ -79,103 +78,120 @@ impl QueryPlan {
     ) -> Response {
-        let response = Arc::new(Mutex::new(Response::builder().build()));
         let root = Path::empty();
 
-        self.root
+        let (value, errors) = self
+            .root
             .execute_recursively(
-                Arc::clone(&response),
                 &root,
                 Arc::clone(&request),
                 Arc::clone(&service_registry),
                 Arc::clone(&schema),
+                &Value::default(),
             )
             .await;
 
-        // TODO: this is not great but there is no other way
-        Arc::try_unwrap(response)
-            .expect("todo: how to prove?")
-            .into_inner()
+        Response::builder().data(value).errors(errors).build()
     }
 }
 
 impl PlanNode {
     fn execute_recursively<'a>(
         &'a self,
-        response: Arc<Mutex<Response>>,
         current_dir: &'a Path,
         request: Arc<Request>,
         service_registry: Arc<dyn ServiceRegistry>,
         schema: Arc<Schema>,
-    ) -> future::BoxFuture<()> {
+        parent_value: &'a Value,
+    ) -> future::BoxFuture<(Value, Vec<Error>)> {
         Box::pin(async move {
             tracing::trace!("Executing plan:\n{:#?}", self);
+            let mut value;
+            let mut errors = Vec::new();
 
             match self {
                 PlanNode::Sequence { nodes } => {
+                    value = parent_value.clone();
                     for node in nodes {
-                        node.execute_recursively(
-                            Arc::clone(&response),
-                            current_dir,
-                            Arc::clone(&request),
-                            Arc::clone(&service_registry),
-                            Arc::clone(&schema),
-                        )
-                        .instrument(tracing::info_span!("sequence"))
-                        .await;
+                        let (v, err) = node
+                            .execute_recursively(
+                                current_dir,
+                                Arc::clone(&request),
+                                Arc::clone(&service_registry),
+                                Arc::clone(&schema),
+                                &value,
+                            )
+                            .instrument(tracing::info_span!("sequence"))
+                            .await;
+                        value.deep_merge(v);
+                        errors.extend(err.into_iter());
                     }
                 }
                 PlanNode::Parallel { nodes } => {
-                    future::join_all(nodes.iter().map(|plan| {
-                        plan.execute_recursively(
-                            Arc::clone(&response),
-                            current_dir,
+                    value = Value::default();
+                    async {
+                        {
+                            let mut stream: FuturesUnordered<_> = nodes
```

Comment:
:information_source: already in scope by future's prelude

```suggestion
                            let mut stream: stream::FuturesUnordered<_> = nodes
```